lib/Dialect/FIRRTL/Transforms/IMConstProp.cpp

```diff
@@ -963,8 +1008,11 @@ void IMConstPropPass::rewriteModuleBody(FModuleOp module) {
       continue;
     }
 
-    // Don't "refold" constants.  TODO: Unique in the module entry block.
-    if (isa<ConstantOp, SpecialConstantOp, InvalidValueOp>(op))
+    // Don't "refold" constants.
+    // If you don't skip here, everything will crash mysteriously and
+    // inconsistently, and asan/ubsan cannot save you.  Here be dragons.
+    if (isa<ConstantOp, SpecialConstantOp, InvalidValueOp>(op) ||
+        op.hasTrait<mlir::OpTrait::ConstantLike>())
```

Comment:
nit: maybe ConstantLike is sufficient here? ConstantOp and SpecialConstantOp have ConstantLike and InvalidValue is now overdefined.

Reply:
Sounds good to me! I'm not sure checking the trait is perfectly legit but short of adding a thing that /does/ mean what we want ("will fold to an attr that will materialize to itself?") this seems right to me-- certainly communicates the intent :+1: .